cmd/genman/gen_kubectl_man.go

```diff
@@ -0,0 +1,177 @@
+		os.Exit(1)
+	}
+
+	docsDir, err := filepath.Abs(docsDir)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	stat, err := os.Stat(docsDir)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "output directory %s does not exist\n", docsDir)
+		os.Exit(1)
+	}
+
+	if !stat.IsDir() {
+		fmt.Fprintf(os.Stderr, "output directory %s is not a directory\n", docsDir)
+		os.Exit(1)
+	}
+	docsDir = docsDir + "/"
+
+	// Set environment variables used by kubectl so the output is consistent,
+	// regardless of where we run.
+	os.Setenv("HOME", "/home/username")
+	kubectl := cmd.NewFactory(nil).NewKubectlCommand(ioutil.Discard)
+	genMarkdown(kubectl, "", docsDir)
+	for _, c := range kubectl.Commands() {
+		genMarkdown(c, "kubectl", docsDir)
+	}
+}
+
+func preamble(out *bytes.Buffer, name, short, long string) {
+	out.WriteString(`% KUBERNETES(1) kubernetes User Manuals
+% Eric Paris
+% Jan 2015
+# NAME
+`)
+	fmt.Fprintf(out, "%s \\- %s\n\n", name, short)
+	fmt.Fprintf(out, "# SYNOPSIS\n")
+	fmt.Fprintf(out, "**%s** [OPTIONS]\n\n", name)
+	fmt.Fprintf(out, "# DESCRIPTION\n")
+	fmt.Fprintf(out, "%s\n\n", long)
+}
+
+func printFlags(out *bytes.Buffer, flags *pflag.FlagSet) {
+	flags.VisitAll(func(flag *pflag.Flag) {
+		format := "**--%s**=%s\n\t%s\n\n"
+		if flag.Value.Type() == "string" {
+			// put quotes on the value
+			format = "**--%s**=%q\n\t%s\n\n"
+		}
+		if len(flag.Shorthand) > 0 {
+			format = "**-%s**, " + format
+		} else {
+			format = "%s" + format
+		}
+		fmt.Fprintf(out, format, flag.Shorthand, flag.Name, flag.DefValue, flag.Usage)
+	})
+}
+
+func printOptions(out *bytes.Buffer, command *cobra.Command) {
+	var flags *pflag.FlagSet
+	if command.HasFlags() {
+		flags = command.Flags()
+	} else if !command.HasParent() && command.HasPersistentFlags() {
+		flags = command.PersistentFlags()
+	}
+	if flags == nil {
+		return
+	}
+	fmt.Fprintf(out, "# OPTIONS\n")
+	printFlags(out, flags)
+	fmt.Fprintf(out, "\n")
+}
+
+func genMarkdown(command *cobra.Command, parent, docsDir string) {
+	dparent := strings.Replace(parent, " ", "-", -1)
+	name := command.Name()
+	dname := name
+	if len(parent) > 0 {
+		dname = dparent + "-" + name
+		name = parent + " " + name
+	}
+
+	out := new(bytes.Buffer)
+	short := command.Short
+	long := command.Long
+	if len(long) == 0 {
+		long = short
+	}
+
+	preamble(out, name, short, long)
+	printOptions(out, command)
+
+	if len(command.Commands()) > 0 || len(parent) > 0 {
+		fmt.Fprintf(out, "# SEE ALSO\n")
+		if len(parent) > 0 {
+			fmt.Fprintf(out, "**%s(1)**, ", dparent)
+		}
+		for _, c := range command.Commands() {
+			fmt.Fprintf(out, "**%s-%s(1)**, ", dname, c.Name())
+			genMarkdown(c, name, docsDir)
+		}
+		fmt.Fprintf(out, "\n")
+	}
+
+	out.WriteString(`
+# HISTORY
+January 2015, Originally compiled by Eric Paris (eparis at redhat dot com) based on the kubernetes source material, but hopefully they have been automatically generated since!
+`)
```

Comment:
Optional/TODO: write the current date here, e.g. "last compiled: MMM DD YYYY", so it gets updated everytime we run `hack/run-gendocs.sh`.